Flaky NS (Redis) test

<!--
Thanks for submitting a bug report. Please fill the template below,
otherwise we will not be able to process this bug report.
-->

#### Summary
<!-- Summarize the problem in a few sentences: -->

There's one test that's consistently failing randomly 😛  

#### Steps to Reproduce
<!-- How can we reproduce the problem? -->

1. Push to a PR
2. Check travis output.
3. Either the `ARCH=amd64` or the `ARCH=386` fails randomly

<!--
Please upload relevant configuration (as .txt).
If you use the command "ttn-lw-stack config", you can redact sensitive config.
-->



#### What do you see now?
<!--
Please paste terminal output, upload logs (as .txt) or upload screenshots.
-->

<img width="536" alt="Travis" src="https://user-images.githubusercontent.com/13186113/59844639-b33df580-935b-11e9-92d6-2d6c0e157c10.png">


```
namespace=networkserver task_id=process_downlink test_name=TestFlow/Redis_registry/Redis_downlink_task_queue/Class_A/OTAA/EU868/1.0.2
--- FAIL: TestFlow (0.00s)
    --- FAIL: TestFlow/Redis_registry/Redis_downlink_task_queue (32.78s)
        --- FAIL: TestFlow/Redis_registry/Redis_downlink_task_queue/Class_A/OTAA/EU868/1.0.2 (32.78s)
            networkserver_flow_test.go:208: Cluster.GetPeer called
            networkserver_flow_test.go:208: ApplicationAccess.ListRights called
            networkserver_flow_test.go:214: Cluster.GetPeer called
            networkserver_flow_test.go:214: ApplicationAccess.ListRights called
            networkserver_flow_test.go:240: Device created
            networkserver_flow_test.go:322: Cluster.GetPeer called
            networkserver_util_test.go:458: Cluster.Auth called
            networkserver_flow_test.go:460: Cluster.GetPeer called
            networkserver_util_test.go:488: Cluster.Auth called
            networkserver_flow_test.go:508: Device successfully joined. DevAddr: C03E4049
            --- FAIL: TestFlow/Redis_registry/Redis_downlink_task_queue/Class_A/OTAA/EU868/1.0.2/uplink (32.75s)
                networkserver_flow_test.go:573: Timed out while waiting for deduplication window to close
            networkserver_flow_test.go:562: HandleUplink returned rpc error: code = Unavailable desc = transport is closing
FAIL
FAIL	go.thethings.network/lorawan-stack/pkg/networkserver	39.564s
```

#### What do you want to see instead?
<!-- Please add some examples or mock-ups if applicable. -->

No failure as for the other build config.

#### Environment
<!--
Your environment: OS/Browser/Gateway/Device/...? Versions? IDs/EUIs?
Paste the output of "ttn-lw-cli version" or "ttn-lw-stack version" if applicable.
-->

Travis, `ttn-lw-stack: v3.0.3`

#### How do you propose to implement this?
<!-- Please think about how this could be fixed. -->

Up to the NS expert @rvolosatovs 

#### Can you do this yourself and submit a Pull Request?
<!-- You can also @mention experts if you need help with this. -->

Nopes

--- a/cmd/ttn-lw-cli/commands/config.go
+++ b/cmd/ttn-lw-cli/commands/config.go
@@ -33,25 +33,37 @@ type Config struct {
 	InputFormat                  string `name:"input-format" description:"Input format"`
 	OutputFormat                 string `name:"output-format" description:"Output format"`
 	AllowUnknownHosts            bool   `name:"allow-unknown-hosts" description:"Allow sending credentials to unknown hosts"`
-	OAuthServerAddress           string `name:"oauth-server-address" description:"OAuth Server Address"`
-	IdentityServerGRPCAddress    string `name:"identity-server-grpc-address" description:"Identity Server Address"`
-	GatewayServerGRPCAddress     string `name:"gateway-server-grpc-address" description:"Gateway Server Address"`
-	NetworkServerGRPCAddress     string `name:"network-server-grpc-address" description:"Network Server Address"`
-	ApplicationServerGRPCAddress string `name:"application-server-grpc-address" description:"Application Server Address"`
-	JoinServerGRPCAddress        string `name:"join-server-grpc-address" description:"Join Server Address"`
+	OAuthServerAddress           string `name:"oauth-server-address" description:"OAuth Server address"`
+	IdentityServerGRPCAddress    string `name:"identity-server-grpc-address" description:"Identity Server address"`
+	GatewayServerEnabled         bool   `name:"gateway-server-enabled" description:"Gateway Server enabled"`
+	GatewayServerGRPCAddress     string `name:"gateway-server-grpc-address" description:"Gateway Server address"`
+	NetworkServerEnabled         bool   `name:"network-server-enabled" description:"Network Server enabled"`
+	NetworkServerGRPCAddress     string `name:"network-server-grpc-address" description:"Network Server address"`
+	ApplicationServerEnabled     bool   `name:"application-server-enabled" description:"Application Server enabled"`
+	ApplicationServerGRPCAddress string `name:"application-server-grpc-address" description:"Application Server address"`
+	JoinServerEnabled            bool   `name:"join-server-enabled" description:"Join Server enabled"`
+	JoinServerGRPCAddress        string `name:"join-server-grpc-address" description:"Join Server address"`
 	Insecure                     bool   `name:"insecure" description:"Connect without TLS"`
 	CA                           string `name:"ca" description:"CA certificate file"`
 }
 
 func (c Config) getHosts() []string {
-	return getHosts(
-		c.OAuthServerAddress,
-		c.IdentityServerGRPCAddress,
-		c.GatewayServerGRPCAddress,
-		c.NetworkServerGRPCAddress,
-		c.ApplicationServerGRPCAddress,
-		c.JoinServerGRPCAddress,
-	)
+	hosts := make([]string, 0, 6)
+	hosts = append(hosts, c.OAuthServerAddress)
+	hosts = append(hosts, c.IdentityServerGRPCAddress)
+	if c.GatewayServerEnabled {
+		hosts = append(hosts, c.GatewayServerGRPCAddress)
+	}
+	if c.NetworkServerEnabled {
+		hosts = append(hosts, c.NetworkServerGRPCAddress)
+	}
+	if c.ApplicationServerEnabled {
+		hosts = append(hosts, c.ApplicationServerGRPCAddress)
+	}
+	if c.JoinServerEnabled {
+		hosts = append(hosts, c.JoinServerGRPCAddress)
+	}
+	return getHosts(hosts...)
 }
 
 // DefaultConfig contains the default config for the ttn-lw-cli binary.
@@ -65,9 +77,13 @@ var DefaultConfig = Config{
 	OutputFormat:                 "json",
 	OAuthServerAddress:           clusterHTTPAddress + "/oauth",
 	IdentityServerGRPCAddress:    clusterGRPCAddress,
+	GatewayServerEnabled:         true,
 	GatewayServerGRPCAddress:     clusterGRPCAddress,
+	NetworkServerEnabled:         true,
 	NetworkServerGRPCAddress:     clusterGRPCAddress,
+	ApplicationServerEnabled:     true,
 	ApplicationServerGRPCAddress: clusterGRPCAddress,
+	JoinServerEnabled:            true,
 	JoinServerGRPCAddress:        clusterGRPCAddress,
 }
 

--- a/cmd/ttn-lw-cli/commands/end_devices.go
+++ b/cmd/ttn-lw-cli/commands/end_devices.go
@@ -141,6 +141,8 @@ func generateDevAddr(netID types.NetID) (types.DevAddr, error) {
 	return devAddr, nil
 }
 
+var errGatewayServerDisabled = errors.DefineFailedPrecondition("gateway_server_disabled", "Gateway Server is disabled")
+
 var (
 	endDevicesCommand = &cobra.Command{
 		Use:     "end-devices",
@@ -152,6 +154,10 @@ var (
 		Short:             "List available frequency plans for end devices",
 		PersistentPreRunE: preRun(),
 		RunE: func(cmd *cobra.Command, args []string) error {
+			if !config.GatewayServerEnabled {
+				return errGatewayServerDisabled
+			}
+
 			baseFrequency, _ := cmd.Flags().GetUint32("base-frequency")
 			ns, err := api.Dial(ctx, config.NetworkServerGRPCAddress)
 			if err != nil {
@@ -309,17 +315,21 @@ var (
 
 			setDefaults, _ := cmd.Flags().GetBool("defaults")
 			if setDefaults {
-				device.NetworkServerAddress = getHost(config.NetworkServerGRPCAddress)
-				device.ApplicationServerAddress = getHost(config.ApplicationServerGRPCAddress)
-				paths = append(paths,
-					"application_server_address",
-					"network_server_address",
-				)
+				if config.NetworkServerEnabled {
+					device.NetworkServerAddress = getHost(config.NetworkServerGRPCAddress)
+					paths = append(paths, "network_server_address")
+				}
+				if config.ApplicationServerEnabled {
+					device.ApplicationServerAddress = getHost(config.ApplicationServerGRPCAddress)
+					paths = append(paths, "application_server_address")
+				}
 			}
 
 			if abp, _ := cmd.Flags().GetBool("abp"); abp {
 				device.SupportsJoin = false
-				paths = append(paths, "supports_join")
+				if config.NetworkServerEnabled {
+					paths = append(paths, "supports_join")
+				}
 				if withSession, _ := cmd.Flags().GetBool("with-session"); withSession {
 					if device.ProvisionerID != "" {
 						return errEndDeviceKeysWithProvisioner
@@ -355,12 +365,16 @@ var (
 				}
 			} else {
 				device.SupportsJoin = true
-				paths = append(paths, "supports_join")
+				if config.NetworkServerEnabled {
+					paths = append(paths, "supports_join")
+				}
 				if setDefaults {
-					device.JoinServerAddress = getHost(config.JoinServerGRPCAddress)
-					paths = append(paths,
-						"join_server_address",
-					)
+					if config.JoinServerEnabled {
+						device.JoinServerAddress = getHost(config.JoinServerGRPCAddress)
+						paths = append(paths,
+							"join_server_address",
+						)
+					}
 				}
 				if withKeys, _ := cmd.Flags().GetBool("with-root-keys"); withKeys {
 					if device.ProvisionerID != "" {
@@ -478,13 +492,13 @@ var (
 
 			isPaths, nsPaths, asPaths, jsPaths := splitEndDeviceSetPaths(device.SupportsJoin, paths...)
 
-			if len(nsPaths) > 0 {
+			if len(nsPaths) > 0 && config.NetworkServerEnabled {
 				isPaths = append(isPaths, "network_server_address")
 			}
-			if len(asPaths) > 0 {
+			if len(asPaths) > 0 && config.ApplicationServerEnabled {
 				isPaths = append(isPaths, "application_server_address")
 			}
-			if len(jsPaths) > 0 {
+			if len(jsPaths) > 0 && config.JoinServerEnabled {
 				isPaths = append(isPaths, "join_server_address")
 			}
 
@@ -744,21 +758,21 @@ var errAddressMismatchEndDevice = errors.DefineAborted("end_device_server_addres
 
 func compareServerAddressesEndDevice(device *ttnpb.EndDevice, config *Config) (nsMismatch, asMismatch, jsMismatch bool) {
 	nsHost, asHost, jsHost := getHost(config.NetworkServerGRPCAddress), getHost(config.ApplicationServerGRPCAddress), getHost(config.JoinServerGRPCAddress)
-	if host := getHost(device.NetworkServerAddress); host != "" && host != nsHost {
+	if host := getHost(device.NetworkServerAddress); config.NetworkServerEnabled && host != "" && host != nsHost {
 		nsMismatch = true
 		logger.WithFields(log.Fields(
 			"configured", nsHost,
 			"registered", host,
 		)).Warn("Registered Network Server address does not match CLI configuration")
 	}
-	if host := getHost(device.ApplicationServerAddress); host != "" && host != asHost {
+	if host := getHost(device.ApplicationServerAddress); config.ApplicationServerEnabled && host != "" && host != asHost {
 		asMismatch = true
 		logger.WithFields(log.Fields(
 			"configured", asHost,
 			"registered", host,
 		)).Warn("Registered Application Server address does not match CLI configuration")
 	}
-	if host := getHost(device.JoinServerAddress); host != "" && host != jsHost {
+	if host := getHost(device.JoinServerAddress); config.JoinServerEnabled && host != "" && host != jsHost {
 		jsMismatch = true
 		logger.WithFields(log.Fields(
 			"configured", jsHost,

--- a/cmd/ttn-lw-cli/commands/end_devices.go
+++ b/cmd/ttn-lw-cli/commands/end_devices.go
@@ -480,9 +480,8 @@ var (
 				return nil
 			}
 			var device ttnpb.EndDevice
-			if ttnpb.HasAnyField(ttnpb.TopLevelFields(paths), "root_keys") {
+			if ttnpb.HasAnyField(paths, setEndDeviceToJS...) {
 				device.SupportsJoin = true
-				paths = append(paths, "supports_join")
 			}
 			if err = util.SetFields(&device, setEndDeviceFlags); err != nil {
 				return err

--- a/cmd/ttn-lw-cli/commands/end_devices.go
+++ b/cmd/ttn-lw-cli/commands/end_devices.go
@@ -153,11 +153,11 @@ var (
 		PersistentPreRunE: preRun(),
 		RunE: func(cmd *cobra.Command, args []string) error {
 			baseFrequency, _ := cmd.Flags().GetUint32("base-frequency")
-			gs, err := api.Dial(ctx, config.NetworkServerGRPCAddress)
+			ns, err := api.Dial(ctx, config.NetworkServerGRPCAddress)
 			if err != nil {
 				return err
 			}
-			res, err := ttnpb.NewConfigurationClient(gs).ListFrequencyPlans(ctx, &ttnpb.ListFrequencyPlansRequest{
+			res, err := ttnpb.NewConfigurationClient(ns).ListFrequencyPlans(ctx, &ttnpb.ListFrequencyPlansRequest{
 				BaseFrequency: baseFrequency,
 			})
 			if err != nil {

--- a/cmd/ttn-lw-cli/commands/end_devices_split.go
+++ b/cmd/ttn-lw-cli/commands/end_devices_split.go
@@ -71,93 +71,104 @@ func splitEndDeviceSetPaths(supportsJoin bool, paths ...string) (is, ns, as, js 
 
 func getEndDevice(ids ttnpb.EndDeviceIdentifiers, nsPaths, asPaths, jsPaths []string, continueOnError bool) (*ttnpb.EndDevice, error) {
 	var res ttnpb.EndDevice
-
 	if len(jsPaths) > 0 {
-		js, err := api.Dial(ctx, config.JoinServerGRPCAddress)
-		if err != nil {
-			if !continueOnError {
-				return nil, err
-			}
-			logger.WithError(err).Error("Could not connect to Join Server")
+		if !config.JoinServerEnabled {
+			logger.WithField("paths", jsPaths).Warn("Join Server disabled but fields specified to get")
 		} else {
-			logger.WithField("paths", jsPaths).Debug("Get EndDevice from Join Server")
-			jsRes, err := ttnpb.NewJsEndDeviceRegistryClient(js).Get(ctx, &ttnpb.GetEndDeviceRequest{
-				EndDeviceIdentifiers: ids,
-				FieldMask:            types.FieldMask{Paths: jsPaths},
-			})
+			js, err := api.Dial(ctx, config.JoinServerGRPCAddress)
 			if err != nil {
 				if !continueOnError {
 					return nil, err
 				}
-				logger.WithError(err).Error("Could not get end device from Join Server")
+				logger.WithError(err).Error("Could not connect to Join Server")
 			} else {
-				res.SetFields(jsRes, ttnpb.AllowedBottomLevelFields(jsPaths, getEndDeviceFromJS)...)
-				if res.CreatedAt.IsZero() || (!jsRes.CreatedAt.IsZero() && jsRes.CreatedAt.Before(res.CreatedAt)) {
-					res.CreatedAt = jsRes.CreatedAt
-				}
-				if jsRes.UpdatedAt.After(res.UpdatedAt) {
-					res.UpdatedAt = jsRes.UpdatedAt
+				logger.WithField("paths", jsPaths).Debug("Get end device from Join Server")
+				jsRes, err := ttnpb.NewJsEndDeviceRegistryClient(js).Get(ctx, &ttnpb.GetEndDeviceRequest{
+					EndDeviceIdentifiers: ids,
+					FieldMask:            types.FieldMask{Paths: jsPaths},
+				})
+				if err != nil {
+					if !continueOnError {
+						return nil, err
+					}
+					logger.WithError(err).Error("Could not get end device from Join Server")
+				} else {
+					res.SetFields(jsRes, ttnpb.AllowedBottomLevelFields(jsPaths, getEndDeviceFromJS)...)
+					if res.CreatedAt.IsZero() || (!jsRes.CreatedAt.IsZero() && jsRes.CreatedAt.Before(res.CreatedAt)) {
+						res.CreatedAt = jsRes.CreatedAt
+					}
+					if jsRes.UpdatedAt.After(res.UpdatedAt) {
+						res.UpdatedAt = jsRes.UpdatedAt
+					}
 				}
 			}
 		}
 	}
 
 	if len(asPaths) > 0 {
-		as, err := api.Dial(ctx, config.ApplicationServerGRPCAddress)
-		if err != nil {
-			if !continueOnError {
-				return nil, err
-			}
-			logger.WithError(err).Error("Could not connect to Application Server")
+		if !config.ApplicationServerEnabled {
+			logger.WithField("paths", asPaths).Warn("Application Server disabled but fields specified to get")
 		} else {
-			logger.WithField("paths", asPaths).Debug("Get EndDevice from Application Server")
-			asRes, err := ttnpb.NewAsEndDeviceRegistryClient(as).Get(ctx, &ttnpb.GetEndDeviceRequest{
-				EndDeviceIdentifiers: ids,
-				FieldMask:            types.FieldMask{Paths: asPaths},
-			})
+			as, err := api.Dial(ctx, config.ApplicationServerGRPCAddress)
 			if err != nil {
 				if !continueOnError {
 					return nil, err
 				}
-				logger.WithError(err).Error("Could not get end device from Application Server")
+				logger.WithError(err).Error("Could not connect to Application Server")
 			} else {
-				res.SetFields(asRes, ttnpb.AllowedBottomLevelFields(asPaths, getEndDeviceFromAS)...)
-				if res.CreatedAt.IsZero() || (!asRes.CreatedAt.IsZero() && asRes.CreatedAt.Before(res.CreatedAt)) {
-					res.CreatedAt = asRes.CreatedAt
-				}
-				if asRes.UpdatedAt.After(res.UpdatedAt) {
-					res.UpdatedAt = asRes.UpdatedAt
+				logger.WithField("paths", asPaths).Debug("Get end device from Application Server")
+				asRes, err := ttnpb.NewAsEndDeviceRegistryClient(as).Get(ctx, &ttnpb.GetEndDeviceRequest{
+					EndDeviceIdentifiers: ids,
+					FieldMask:            types.FieldMask{Paths: asPaths},
+				})
+				if err != nil {
+					if !continueOnError {
+						return nil, err
+					}
+					logger.WithError(err).Error("Could not get end device from Application Server")
+				} else {
+					res.SetFields(asRes, ttnpb.AllowedBottomLevelFields(asPaths, getEndDeviceFromAS)...)
+					if res.CreatedAt.IsZero() || (!asRes.CreatedAt.IsZero() && asRes.CreatedAt.Before(res.CreatedAt)) {
+						res.CreatedAt = asRes.CreatedAt
+					}
+					if asRes.UpdatedAt.After(res.UpdatedAt) {
+						res.UpdatedAt = asRes.UpdatedAt
+					}
 				}
 			}
 		}
 	}
 
 	if len(nsPaths) > 0 {
-		ns, err := api.Dial(ctx, config.NetworkServerGRPCAddress)
-		if err != nil {
-			if !continueOnError {
-				return nil, err
-			}
-			logger.WithError(err).Error("Could not connect to Network Server")
+		if !config.NetworkServerEnabled {
+			logger.WithField("paths", nsPaths).Warn("Network Server disabled but fields specified to get")
 		} else {
-			logger.WithField("paths", nsPaths).Debug("Get EndDevice from Network Server")
-			nsRes, err := ttnpb.NewNsEndDeviceRegistryClient(ns).Get(ctx, &ttnpb.GetEndDeviceRequest{
-				EndDeviceIdentifiers: ids,
-				FieldMask:            types.FieldMask{Paths: nsPaths},
-			})
+			ns, err := api.Dial(ctx, config.NetworkServerGRPCAddress)
 			if err != nil {
 				if !continueOnError {
 					return nil, err
 				}
-				logger.WithError(err).Error("Could not get end device from Network Server")
+				logger.WithError(err).Error("Could not connect to Network Server")
 			} else {
-				res.SetFields(nsRes, "ids.dev_addr")
-				res.SetFields(nsRes, ttnpb.AllowedBottomLevelFields(nsPaths, getEndDeviceFromNS)...)
-				if res.CreatedAt.IsZero() || (!nsRes.CreatedAt.IsZero() && nsRes.CreatedAt.Before(res.CreatedAt)) {
-					res.CreatedAt = nsRes.CreatedAt
-				}
-				if nsRes.UpdatedAt.After(res.UpdatedAt) {
-					res.UpdatedAt = nsRes.UpdatedAt
+				logger.WithField("paths", nsPaths).Debug("Get end device from Network Server")
+				nsRes, err := ttnpb.NewNsEndDeviceRegistryClient(ns).Get(ctx, &ttnpb.GetEndDeviceRequest{
+					EndDeviceIdentifiers: ids,
+					FieldMask:            types.FieldMask{Paths: nsPaths},
+				})
+				if err != nil {
+					if !continueOnError {
+						return nil, err
+					}
+					logger.WithError(err).Error("Could not get end device from Network Server")
+				} else {
+					res.SetFields(nsRes, "ids.dev_addr")
+					res.SetFields(nsRes, ttnpb.AllowedBottomLevelFields(nsPaths, getEndDeviceFromNS)...)
+					if res.CreatedAt.IsZero() || (!nsRes.CreatedAt.IsZero() && nsRes.CreatedAt.Before(res.CreatedAt)) {
+						res.CreatedAt = nsRes.CreatedAt
+					}
+					if nsRes.UpdatedAt.After(res.UpdatedAt) {
+						res.UpdatedAt = nsRes.UpdatedAt
+					}
 				}
 			}
 		}
@@ -176,7 +187,7 @@ func setEndDevice(device *ttnpb.EndDevice, isPaths, nsPaths, asPaths, jsPaths []
 			return nil, err
 		}
 		var isDevice ttnpb.EndDevice
-		logger.WithField("paths", isPaths).Debug("Set EndDevice on Identity Server")
+		logger.WithField("paths", isPaths).Debug("Set end device on Identity Server")
 		isDevice.SetFields(device, append(isPaths, "ids")...)
 		isRes, err := ttnpb.NewEndDeviceRegistryClient(is).Update(ctx, &ttnpb.UpdateEndDeviceRequest{
 			EndDevice: isDevice,
@@ -194,13 +205,15 @@ func setEndDevice(device *ttnpb.EndDevice, isPaths, nsPaths, asPaths, jsPaths []
 		}
 	}
 
-	if len(jsPaths) > 0 {
+	if len(jsPaths) > 0 && !config.JoinServerEnabled {
+		logger.WithField("paths", jsPaths).Warn("Join Server disabled but fields specified to set")
+	} else if len(jsPaths) > 0 {
 		js, err := api.Dial(ctx, config.JoinServerGRPCAddress)
 		if err != nil {
 			return nil, err
 		}
 		var jsDevice ttnpb.EndDevice
-		logger.WithField("paths", jsPaths).Debug("Set EndDevice on Join Server")
+		logger.WithField("paths", jsPaths).Debug("Set end device on Join Server")
 		jsDevice.SetFields(device, append(jsPaths, "ids")...)
 		jsRes, err := ttnpb.NewJsEndDeviceRegistryClient(js).Set(ctx, &ttnpb.SetEndDeviceRequest{
 			EndDevice: jsDevice,
@@ -218,13 +231,15 @@ func setEndDevice(device *ttnpb.EndDevice, isPaths, nsPaths, asPaths, jsPaths []
 		}
 	}
 
-	if len(nsPaths) > 0 || isCreate {
+	if len(nsPaths) > 0 && !config.NetworkServerEnabled {
+		logger.WithField("paths", nsPaths).Warn("Network Server disabled but fields specified to set")
+	} else if (len(nsPaths) > 0 || isCreate) && config.NetworkServerEnabled {
 		ns, err := api.Dial(ctx, config.NetworkServerGRPCAddress)
 		if err != nil {
 			return nil, err
 		}
 		var nsDevice ttnpb.EndDevice
-		logger.WithField("paths", nsPaths).Debug("Set EndDevice on Network Server")
+		logger.WithField("paths", nsPaths).Debug("Set end device on Network Server")
 		nsDevice.SetFields(device, append(nsPaths, "ids")...)
 		nsRes, err := ttnpb.NewNsEndDeviceRegistryClient(ns).Set(ctx, &ttnpb.SetEndDeviceRequest{
 			EndDevice: nsDevice,
@@ -242,13 +257,15 @@ func setEndDevice(device *ttnpb.EndDevice, isPaths, nsPaths, asPaths, jsPaths []
 		}
 	}
 
-	if len(asPaths) > 0 || isCreate {
+	if len(asPaths) > 0 && !config.ApplicationServerEnabled {
+		logger.WithField("paths", asPaths).Warn("Application Server disabled but fields specified to set")
+	} else if (len(asPaths) > 0 || isCreate) && config.ApplicationServerEnabled {
 		as, err := api.Dial(ctx, config.ApplicationServerGRPCAddress)
 		if err != nil {
 			return nil, err
 		}
 		var asDevice ttnpb.EndDevice
-		logger.WithField("paths", asPaths).Debug("Set EndDevice on Application Server")
+		logger.WithField("paths", asPaths).Debug("Set end device on Application Server")
 		asDevice.SetFields(device, append(asPaths, "ids")...)
 		asRes, err := ttnpb.NewAsEndDeviceRegistryClient(as).Set(ctx, &ttnpb.SetEndDeviceRequest{
 			EndDevice: asDevice,
@@ -270,38 +287,44 @@ func setEndDevice(device *ttnpb.EndDevice, isPaths, nsPaths, asPaths, jsPaths []
 }
 
 func deleteEndDevice(ctx context.Context, devID *ttnpb.EndDeviceIdentifiers) error {
-	as, err := api.Dial(ctx, config.ApplicationServerGRPCAddress)
-	if err != nil {
-		return err
-	}
-	_, err = ttnpb.NewAsEndDeviceRegistryClient(as).Delete(ctx, devID)
-	if errors.IsNotFound(err) {
-		logger.WithError(err).Error("Could not delete end device from Application Server")
-	} else if err != nil {
-		return err
-	}
-
-	ns, err := api.Dial(ctx, config.NetworkServerGRPCAddress)
-	if err != nil {
-		return err
-	}
-	_, err = ttnpb.NewNsEndDeviceRegistryClient(ns).Delete(ctx, devID)
-	if errors.IsNotFound(err) {
-		logger.WithError(err).Error("Could not delete end device from Network Server")
-	} else if err != nil {
-		return err
-	}
-
-	if devID.JoinEUI != nil && devID.DevEUI != nil {
-		js, err := api.Dial(ctx, config.JoinServerGRPCAddress)
+	if config.ApplicationServerEnabled {
+		as, err := api.Dial(ctx, config.ApplicationServerGRPCAddress)
 		if err != nil {
 			return err
 		}
-		_, err = ttnpb.NewJsEndDeviceRegistryClient(js).Delete(ctx, devID)
+		_, err = ttnpb.NewAsEndDeviceRegistryClient(as).Delete(ctx, devID)
 		if errors.IsNotFound(err) {
-			logger.WithError(err).Error("Could not delete end device from Join Server")
+			logger.WithError(err).Error("Could not delete end device from Application Server")
 		} else if err != nil {
 			return err
+		}
+	}
+
+	if config.NetworkServerEnabled {
+		ns, err := api.Dial(ctx, config.NetworkServerGRPCAddress)
+		if err != nil {
+			return err
+		}
+		_, err = ttnpb.NewNsEndDeviceRegistryClient(ns).Delete(ctx, devID)
+		if errors.IsNotFound(err) {
+			logger.WithError(err).Error("Could not delete end device from Network Server")
+		} else if err != nil {
+			return err
+		}
+	}
+
+	if config.JoinServerEnabled {
+		if devID.JoinEUI != nil && devID.DevEUI != nil {
+			js, err := api.Dial(ctx, config.JoinServerGRPCAddress)
+			if err != nil {
+				return err
+			}
+			_, err = ttnpb.NewJsEndDeviceRegistryClient(js).Delete(ctx, devID)
+			if errors.IsNotFound(err) {
+				logger.WithError(err).Error("Could not delete end device from Join Server")
+			} else if err != nil {
+				return err
+			}
 		}
 	}
 

--- a/cmd/ttn-lw-cli/commands/events.go
+++ b/cmd/ttn-lw-cli/commands/events.go
@@ -34,10 +34,18 @@ var eventsCommand = &cobra.Command{
 
 		addresses := make(map[string]bool)
 		addresses[config.IdentityServerGRPCAddress] = true
-		addresses[config.GatewayServerGRPCAddress] = true
-		addresses[config.NetworkServerGRPCAddress] = true
-		addresses[config.ApplicationServerGRPCAddress] = true
-		addresses[config.JoinServerGRPCAddress] = true
+		if config.GatewayServerEnabled {
+			addresses[config.GatewayServerGRPCAddress] = true
+		}
+		if config.NetworkServerEnabled {
+			addresses[config.NetworkServerGRPCAddress] = true
+		}
+		if config.ApplicationServerEnabled {
+			addresses[config.ApplicationServerGRPCAddress] = true
+		}
+		if config.JoinServerEnabled {
+			addresses[config.JoinServerGRPCAddress] = true
+		}
 
 		ids := getCombinedIdentifiers(cmd.Flags()).GetEntityIdentifiers()
 		if len(ids) == 0 {

--- a/config/messages.json
+++ b/config/messages.json
@@ -1196,6 +1196,15 @@
       "file": "gateways.go"
     }
   },
+  "error:cmd/ttn-lw-cli/commands:gateway_server_disabled": {
+    "translations": {
+      "en": "Gateway Server is disabled"
+    },
+    "description": {
+      "package": "cmd/ttn-lw-cli/commands",
+      "file": "end_devices.go"
+    }
+  },
   "error:cmd/ttn-lw-cli/commands:inconsistent_end_device_eui": {
     "translations": {
       "en": "given end device EUIs do not match registered EUIs"

--- a/pkg/networkserver/networkserver_flow_test.go
+++ b/pkg/networkserver/networkserver_flow_test.go
@@ -800,7 +800,7 @@ func TestFlow(t *testing.T) {
 					}
 
 					ctx := test.ContextWithT(test.Context(), t)
-					ctx, cancel := context.WithTimeout(ctx, (1<<12)*test.Delay)
+					ctx, cancel := context.WithTimeout(ctx, (1<<13)*test.Delay)
 					defer cancel()
 					handleFlowTest(ctx, reg, tq)
 				})

